Gil swerve

goooooooooooooooooose opened the pull request.

goooooooooooooooooose committed:
created SwerveModule.java

goooooooooooooooooose committed:
created Drivetrain.java

goooooooooooooooooose committed:
created Drivetrain.java

goooooooooooooooooose committed:
created Drive.java

goooooooooooooooooose committed:
created SwerveModule.java

goooooooooooooooooose committed:
created SwerveModuleHolder.java

goooooooooooooooooose committed:
created SwerveModuleHolder.java

Spikes-2212 commented on src/main/java/frc/robot/RobotMap.java:
```diff
@@ -4,6 +4,21 @@ public class RobotMap {
+        int FRONT_RIGHT_ABSOLUTE_ENCODER = -1;
+        int BACK_LEFT_ABSOLUTE_ENCODER = -1;
+        int BACK_RIGHT_ABSOLUTE_ENCODER = -1;
+
```
remove

Spikes-2212 commented on src/main/java/frc/robot/commands/Drive.java:
```diff
@@ -0,0 +1,36 @@
+
+import java.util.function.Supplier;
+
+public class Drive extends Command {
```
classes begin with a blank line

Spikes-2212 commented on src/main/java/frc/robot/commands/Drive.java:
```diff
@@ -0,0 +1,36 @@
+    private final boolean fieldRelative;
+    private final boolean usePID;
+
+    public Drive( Drivetrain drivetrain, Supplier<Double> xSpeed, Supplier<Double> ySpeed, Supplier<Double> rotationsSpeed, boolean fieldRelative, boolean usePID) {
```
line break

Spikes-2212 commented on src/main/java/frc/robot/commands/Drive.java:
```diff
@@ -0,0 +1,36 @@
+
+    @Override
+    public void execute() {
+        double xSpeed = this.xSpeed.get();
```
no point doing this if not limiting acceleration

Spikes-2212 commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,83 @@
+import edu.wpi.first.math.kinematics.SwerveModuleState;
+import frc.robot.subsystems.SwerveModule;
+
+public class Drivetrain extends DashboardedSubsystem {
```
blank line

Spikes-2212 commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,83 @@
+import frc.robot.subsystems.SwerveModule;
+
+public class Drivetrain extends DashboardedSubsystem {
+    private static final double TRACK_WIDTH = 0.61;
```
set to -1

Spikes-2212 commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,83 @@
+
+public class Drivetrain extends DashboardedSubsystem {
+    private static final double TRACK_WIDTH = 0.61;
+    private static final double TRACK_LENGTH = 0.58;
```
ditto

Spikes-2212 commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,83 @@
+
+    @Override
+    public void configureDashboard() {
+
```
remove

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,175 @@
+    private static final double SECONDS_IN_MINUTE = 60;
+    private static final double ABSOLUTE_POSITION = 1;
+
+
```
remove

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,175 @@
+
+    public static final double MAX_SPEED = 4;
+    public static final double MIN_SPEED = 0.4;
+    public static final double MAX_TURN = 3;
```
blank line after this

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,175 @@
+        configureAbsoluteEncoder();
+    }
+
+    private void createDriveConfig() {
```
this function should be merged with configureDriveController()

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,175 @@
+    }
+
+    private void createDriveConfig() {
+        TalonFXConfiguration config = new TalonFXConfiguration();
```
this is never applied into the motor controller

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,175 @@
+    public void configureTurnController() {
+        ClosedLoopConfig turnClosedLoopConfig = new ClosedLoopConfig();
+        turnClosedLoopConfig.pid(turnPIDSettings.getkP(), turnPIDSettings.getkI(), turnPIDSettings.getkD());
+        turnEncoderConfig.inverted(cancoderInverted);
```
ditto

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,175 @@
+    }
+
+    private void setSpeed(double speed, boolean usePID) {
+        if (usePID) {
```
feed forward?

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,175 @@
+        } else driveMotor.set(speed / MAX_SPEED);
+    }
+
+    private void setAngle(double angle) {
```
feed forward?

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModuleHolder.java:
```diff
@@ -0,0 +1,93 @@
+    private static final double BACK_LEFT_OFFSET = -1;
+    private static final double BACK_RIGHT_OFFSET = -1;
+
+    private static final PIDSettings drivePIDSetting = namespace.addPIDNamespace("drive",
```
drivePIDSettings*

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModuleHolder.java:
```diff
@@ -0,0 +1,93 @@
+
+    private static final PIDSettings drivePIDSetting = namespace.addPIDNamespace("drive",
+            new PIDSettings(-1, -1, -1, -1, -1));
+    private static final PIDSettings turnPIDSetting = namespace.addPIDNamespace("turn",
```
settings

goooooooooooooooooose committed:
removed blank lines.
added blank lines.
added limit speed accelerations in Drive.java to xSpeed, ySpeed and rotationSpeed.
changed track width and track length to equal -1 in Drivetrain.java.
merged create configuration functions with their configure controllers counterpart.
applied configurations to their motor controllers.
added feed forward configuration function.
added feed forward to set angle.
added s to drivePIDSetting and turnPIDSetting in SwerveModuleHolder.java.

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,192 @@
+
+    public static final double MAX_SPEED = 4;
+    public static final double MIN_SPEED = 0.4;
+    public static final double MAX_TURN = 3;
```
MAX_TURN_SPEED

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,192 @@
+    private static final double WHEEL_DIAMETER = 4 * 0.0254;
+    private static final double DEGREES_IN_ROTATIONS = 360;
+    private static final double SECONDS_IN_MINUTE = 60;
+    private static final double ABSOLUTE_POSITION = 1;
```
?

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,192 @@
+        turnEncoderConfig.inverted(cancoderInverted);
+        sparkConfig.apply(turnClosedLoopConfig);
+        turnMotor.configure(sparkConfig, SparkBase.ResetMode.kNoResetSafeParameters, SparkBase.PersistMode.kNoPersistParameters);
+
```
remove

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,192 @@
+
+    private SwerveModuleState optimize(SwerveModuleState state, double currentAngle) {
+        double desiredAngle = normalizeAngleRelativeToEncoder(currentAngle, state.angle.getDegrees());
+        while (Math.abs(desiredAngle - currentAngle) > 90) {
```
magic number

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,192 @@
+        double desiredAngle = normalizeAngleRelativeToEncoder(currentAngle, state.angle.getDegrees());
+        while (Math.abs(desiredAngle - currentAngle) > 90) {
+            if (desiredAngle - currentAngle > 0) {
+                desiredAngle -= 180;
```
ditto

goooooooooooooooooose committed:
renamed MAX_TURN to MAX_TURN_SPEED

goooooooooooooooooose committed:
removed blank line

goooooooooooooooooose committed:
unmagiced

goooooooooooooooooose committed:
unmagiced 2.0

goooooooooooooooooose committed:
renamed ABSOLUTE_POSITION to ABSOLUTE_POSITION_DISCONTINUITY_POINT

Spikes-2212 commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,83 @@
+    private static final double TRACK_WIDTH = -1;
+    private static final double TRACK_LENGTH = -1;
+
+    public static final Translation2d CENTER_OF_ROBOT = new Translation2d(0, 0);
```
these should be private

Spikes-2212 commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,83 @@
+    private final SwerveModule frontRight;
+    private final SwerveModule backLeft;
+    private final SwerveModule backRight;
+    private final SwerveDriveKinematics kinematics;
```
put this below the gyro field and add a blank space

Spikes-2212 commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,83 @@
+        this.frontRight = frontRight;
+        this.backLeft = backLeft;
+        this.backRight = backRight;
+        this.kinematics = new SwerveDriveKinematics(FRONT_LEFT_WHEEL_POSITION,
```
no reason to write write "this"

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,194 @@
+
+    public static final double MAX_SPEED = 4;
+    public static final double MIN_SPEED = 0.4;
+    public static final double MAX_TURN_SPEED = 3;
```
this should be in Drivetrain.java

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,194 @@
+
+    private static final double DRIVE_GEAR_RATIO = 1 / 6.12;
+    private static final double TURN_GEAR_RATIO = 1 / 12.8;
+    private static final double WHEEL_DIAMETER = 4 * 0.0254;
```
seperate this into 2 different constants:
WHEEL_DIAMETER_INCHES
INCHES_TO_METERS

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,194 @@
+    private static final double WHEEL_DIAMETER = 4 * 0.0254;
+    private static final double DEGREES_IN_ROTATIONS = 360;
+    private static final double SECONDS_IN_MINUTE = 60;
+    private static final double ABSOLUTE_POSITION_DISCONTINUITY_POINT = 1;
```
put this below DEGREES_TO_FLIP

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,194 @@
+    private final boolean driveInverted;
+    private final double offset;
+
+    private final RelativeEncoder turnEncoder;
```
put this with the configs below the pid and ff settings

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,194 @@
+        this.cancoderInverted = cancoderInverted;
+        this.driveInverted = driveInverted;
+        this.offset = offset;
+        this.turnEncoder = turnMotor.getEncoder();
```
ditto

Spikes-2212 commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,83 @@
+    private final SwerveDriveKinematics kinematics;
+    private final AHRS gyro;
+
+    public Drivetrain(SwerveModule frontLeft, SwerveModule frontRight, SwerveModule backLeft,
```
turn this into a singleton class

goooooooooooooooooose committed:
made Translation2d fields private
moved gyro ahead of SwerveModules
removes this from kinematics

goooooooooooooooooose committed:
max and min speeds moved from SwerveModule.java to Drivetrain.java

goooooooooooooooooose committed:
separated WHEEL_DIAMETER into WHEEL_DIAMETER_INCHES and INCHES_TO_METERS

goooooooooooooooooose committed:
put ABSOLUTE_POSITION_DISCONTINUITY_POINT below DEGREES_TO_FLIP

goooooooooooooooooose committed:
put turnEncoder below turnFeedForwardController

goooooooooooooooooose committed:
ditto

goooooooooooooooooose committed:
turned Drivetrain.java into a singleton class

Spikes-2212 commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,99 @@
+
+    private static final String NAMESPACE_NAME = "drivetrain";
+
+    private final AHRS gyro;
```
put this field right below the modules

Spikes-2212 commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,99 @@
+        if (instance == null) {
+            instance = new Drivetrain(SwerveModuleHolder.getFrontLeft(), SwerveModuleHolder.getFrontRight(),
+                    SwerveModuleHolder.getBackLeft(), SwerveModuleHolder.getBackRight(),
+                    new AHRS(AHRS.NavXComType.kI2C));
```
make sure this is correct

Spikes-2212 commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,99 @@
+        this.frontRight = frontRight;
+        this.backLeft = backLeft;
+        this.backRight = backRight;
+        kinematics = new SwerveDriveKinematics(FRONT_LEFT_WHEEL_POSITION,
```
swap with the gyro initializer

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,197 @@
+    private final boolean driveInverted;
+    private final double offset;
+
+
```
remove

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,197 @@
+    private final FeedForwardSettings driveFeedForwardSettings;
+    private final FeedForwardSettings turnFeedForwardSettings;
+    private final FeedForwardController turnFeedForwardController;
+    private final RelativeEncoder turnEncoder;
```
seperate from control loop related fields

goooooooooooooooooose committed:
committing just so I can make changes in Swerve

goooooooooooooooooose committed:
committing just so I can make changes in Swerve

goooooooooooooooooose committed:
moved the gyro field right below the modules

goooooooooooooooooose committed:
swapped the gyro initializer with the kinematics initializer

goooooooooooooooooose committed:
removed blank line

goooooooooooooooooose committed:
removed blank line

goooooooooooooooooose committed:
separated the turnEncoder from the control loop related fields

goooooooooooooooooose committed:
separated the turnEncoder from the control loop related fields

goooooooooooooooooose committed:
added a todo to check if new AHRS is correct as well as adding gyro angle to the dashboard

goooooooooooooooooose committed:
added a todo to check if new AHRS is correct as well as adding gyro angle to the dashboard

goooooooooooooooooose committed:
committing in order to change branch

goooooooooooooooooose committed:
committing in order to change branch

tuval88 commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,101 @@
+
+    private static final String NAMESPACE_NAME = "drivetrain";
+
+
```
delete empty line

tuval88 commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,101 @@
+            instance = new Drivetrain(SwerveModuleHolder.getFrontLeft(), SwerveModuleHolder.getFrontRight(),
+                    SwerveModuleHolder.getBackLeft(), SwerveModuleHolder.getBackRight(),
+                    //TODO check in shuffleboard if this is correct
+                    new AHRS(AHRS.NavXComType.kI2C));
```
change to kMXP_SPI as it is more likely to be the one we need

tuval88 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,197 @@
+    }
+
+    public void configureTurnController() {
+        turnEncoderConfig.positionConversionFactor(DRIVE_GEAR_RATIO * DEGREES_IN_ROTATIONS);
```
this is using the wrong gear ratio

tuval88 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,197 @@
+        turnEncoderConfig.velocityConversionFactor((TURN_GEAR_RATIO * DEGREES_IN_ROTATIONS) / SECONDS_IN_MINUTE);
+        ClosedLoopConfig turnClosedLoopConfig = new ClosedLoopConfig();
+        turnClosedLoopConfig.pid(turnPIDSettings.getkP(), turnPIDSettings.getkI(), turnPIDSettings.getkD());
+        turnEncoderConfig.inverted(cancoderInverted);
```
this will crash upon being applied to the motor controller. change this to invert the motor itself rather than the encoder.

tuval88 replied:
this config is never applied to the motor

tuval88 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,197 @@
+        configureTurnController();
+        turnConfigureFF();
+        double feedForward = turnFeedForwardController.calculate(angle);
+        turnMotor.getClosedLoopController().setReference(angle, SparkBase.ControlType.kVelocity, ClosedLoopSlot.kSlot0,
```
this is using the wrong control type

tuval88 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,197 @@
+                InvertedValue.Clockwise_Positive : InvertedValue.CounterClockwise_Positive));
+    }
+
+    public void configureTurnController() {
```
this is being called periodically, but includes configurations which should only happen once. split into two functions, one called periodically and one that is only run once.

tuval88 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,197 @@
+        configureAbsoluteEncoder();
+    }
+
+    public void configureDriveController() {
```
ditto

tuval88 commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,101 @@
+public class Drivetrain extends DashboardedSubsystem {
+
+    public static final double MAX_SPEED = 4;
+    public static final double MIN_SPEED = 0.4;
```
this number seems suspiciously large

tuval88 commented on src/main/java/frc/robot/subsystems/SwerveModuleHolder.java:
```diff
@@ -0,0 +1,93 @@
+    private static final FeedForwardSettings driveFeedForwardSettings = namespace.addFeedForwardNamespace(
+            "drive", new FeedForwardSettings(-1, -1, -1, -1));
+    private static final FeedForwardSettings turnFeedForwardSettings = namespace.addFeedForwardNamespace(
+            "turn", new FeedForwardSettings(-1, -1, -1, -1));
```
don't use -1 as a default value for PID/FF settings, use 0.

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,197 @@
+    private static final double DEGREES_TO_FLIP = 180;
+    private static final double ABSOLUTE_POSITION_DISCONTINUITY_POINT = 1;
+
+    private final TalonFX driveMotor;
```
TalonFXWrapper

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,197 @@
+    private static final double ABSOLUTE_POSITION_DISCONTINUITY_POINT = 1;
+
+    private final TalonFX driveMotor;
+    private final SparkMax turnMotor;
```
SparkWrapper

goooooooooooooooooose committed:
added path planner

goooooooooooooooooose committed:
added path planner

goooooooooooooooooose committed:
deleted extra line

goooooooooooooooooose committed:
deleted extra line

goooooooooooooooooose committed:
changed navx port

goooooooooooooooooose committed:
changed navx port

goooooooooooooooooose committed:
changed gear ratio from drive to turn

goooooooooooooooooose committed:
changed gear ratio from drive to turn

goooooooooooooooooose committed:
updated SpikesLib2.json

goooooooooooooooooose committed:
updated SpikesLib2.json

goooooooooooooooooose committed:
added wait correct parameters

goooooooooooooooooose committed:
added wait correct parameters

goooooooooooooooooose committed:
replaced motor controllers with wrappers

goooooooooooooooooose committed:
replaced motor controllers with wrappers

goooooooooooooooooose committed:
made set angle efficient

goooooooooooooooooose committed:
set default value for pid and ff to 0

goooooooooooooooooose committed:
set default value for pid and ff to 0

goooooooooooooooooose committed:
reduced min speed

goooooooooooooooooose committed:
reduced min speed

TurtleMinecraft commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,100 @@
+    public static final double MIN_SPEED = 0.05;
+    public static final double MAX_TURN_SPEED = 3;
+
+    private static final double TRACK_WIDTH = -1;
```
ask mechanics for the width

TurtleMinecraft commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,100 @@
+    public static final double MAX_TURN_SPEED = 3;
+
+    private static final double TRACK_WIDTH = -1;
+    private static final double TRACK_LENGTH = -1;
```
ditto

TurtleMinecraft commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,100 @@
+        if (instance == null) {
+            instance = new Drivetrain(SwerveModuleHolder.getFrontLeft(), SwerveModuleHolder.getFrontRight(),
+                    SwerveModuleHolder.getBackLeft(), SwerveModuleHolder.getBackRight(),
+                    //TODO check in shuffleboard if this is correct
```
delete

TurtleMinecraft commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,100 @@
+        backRight.resetRelativeEncoder();
+    }
+
+
```
remove

TurtleMinecraft commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,100 @@
+
+    @Override
+    public void configureDashboard() {
+        namespace.putNumber("gyro", gyro::getAngle);
```
what about the gyro?

TurtleMinecraft commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,161 @@
+    private final FeedForwardController turnFeedForwardController;
+
+    private final EncoderConfig turnEncoderConfig;
+    private final SparkMaxConfig sparkConfig;
```
is this used? if not then delete

TurtleMinecraft commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,161 @@
+    private final FeedForwardSettings turnFeedForwardSettings;
+    private final FeedForwardController turnFeedForwardController;
+
+    private final EncoderConfig turnEncoderConfig;
```
ditto

TurtleMinecraft commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,161 @@
+
+    private final EncoderConfig turnEncoderConfig;
+    private final SparkMaxConfig sparkConfig;
+    private final MotorOutputConfigs MotorOutput;
```
ditto

TurtleMinecraft commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,161 @@
+    private final PIDSettings turnPIDSettings;
+    private final FeedForwardSettings driveFeedForwardSettings;
+    private final FeedForwardSettings turnFeedForwardSettings;
+    private final FeedForwardController turnFeedForwardController;
```
![image](https://github.com/user-attachments/assets/b1cd6ec8-58bb-4a09-87ef-def3e0a316f6)

TurtleMinecraft commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,161 @@
+        //@TODO check if this is correct
+        driveMotor.setEncoderConversionFactor(DRIVE_GEAR_RATIO * WHEEL_DIAMETER_INCHES
+                * INCHES_TO_METERS * Math.PI);
+        driveMotor.configurePID(drivePIDSettings);
+        driveMotor.configureFF(driveFeedForwardSettings);
```
no reason to do this

TurtleMinecraft commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,161 @@
+    public void configureTurnController() {
+        turnMotor.setPositionConversionFactor(TURN_GEAR_RATIO * DEGREES_IN_ROTATIONS);
+        turnMotor.setVelocityConversionFactor((TURN_GEAR_RATIO * DEGREES_IN_ROTATIONS) / SECONDS_IN_MINUTE);
+        turnMotor.configurePID(turnPIDSettings);
```
ditto

TurtleMinecraft commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,161 @@
+        turnMotor.setVelocityConversionFactor((TURN_GEAR_RATIO * DEGREES_IN_ROTATIONS) / SECONDS_IN_MINUTE);
+        turnMotor.configurePID(turnPIDSettings);
+        turnMotor.setInverted(cancoderInverted);
+        turnMotor.configureFF(turnFeedForwardSettings);
```
ditto

TurtleMinecraft commented on vendordeps/SpikesLib2.json:
```diff
@@ -0,0 +1,21 @@
+      "version": "v5.1.0"
+    }
+  ]
+}
```
classes end with a blank line

goooooooooooooooooose committed:
deleted todo

goooooooooooooooooose committed:
deleted todo

goooooooooooooooooose committed:
deleted extra line

goooooooooooooooooose committed:
deleted extra line

goooooooooooooooooose committed:
specified what about the gyro

goooooooooooooooooose committed:
specified what about the gyro

goooooooooooooooooose committed:
removed unnecessary fields

goooooooooooooooooose committed:
removed unnecessary fields

goooooooooooooooooose committed:
removed unnecessary pid configurations

goooooooooooooooooose committed:
removed unnecessary pid configurations

goooooooooooooooooose committed:
added blank line at the end of class

goooooooooooooooooose committed:
added blank line at the end of class

TurtleMinecraft commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,98 @@
+        backRight.stop();
+    }
+
+    public void resetRelativeEncoder() {
```
resetRelativeEncoder*s*

TurtleMinecraft commented on src/main/java/frc/robot/commands/Drive.java:
```diff
@@ -0,0 +1,49 @@
+        this.rotationsSpeed = rotationsSpeed;
+        this.fieldRelative = fieldRelative;
+        this.usePID = usePID;
+        this.xLimiter = new SlewRateLimiter(DRIVE_ACCELERATION_LIMIT);
+        this.yLimiter = new SlewRateLimiter(DRIVE_ACCELERATION_LIMIT);
+        this.rotationLimiter = new SlewRateLimiter(TURN_ACCELERATION_LIMIT);
```
unnecessary "this" call

TurtleMinecraft commented on src/main/java/frc/robot/commands/Drive.java:
```diff
@@ -0,0 +1,49 @@
+    private final SlewRateLimiter yLimiter;
+    private final SlewRateLimiter rotationLimiter;
+
+    public Drive(Drivetrain drivetrain, Supplier<Double> xSpeed, Supplier<Double> ySpeed,
```
this command should require the drivetrain

goooooooooooooooooose replied:
what do you meannnn

TurtleMinecraft replied:
addRequirements(drivetrain);

goooooooooooooooooose committed:
renamed reset relative encoders func

goooooooooooooooooose committed:
renamed reset relative encoders func

goooooooooooooooooose committed:
fixed unnecessary this call

goooooooooooooooooose committed:
fixed unnecessary this call

goooooooooooooooooose committed:
added track variables in cm

goooooooooooooooooose committed:
added track variables in cm

goooooooooooooooooose committed:
removed unnecessary this call

goooooooooooooooooose committed:
removed unnecessary this call

tuval88 commented on src/main/java/frc/robot/commands/Drive.java:
```diff
@@ -0,0 +1,50 @@
+    private final Drivetrain drivetrain;
+    private final Supplier<Double> xSpeed;
+    private final Supplier<Double> ySpeed;
+    private final Supplier<Double> rotationsSpeed;
```
rename to rotationSpeed

tuval88 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,148 @@
+            if (desiredAngle - currentAngle > 0) {
+                desiredAngle -= DEGREES_TO_FLIP;
+            }
+            else {
```
merge into one line

tuval88 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,148 @@
+    }
+
+    public void set(SwerveModuleState state, boolean usePID) {
+        if (state.speedMetersPerSecond < Drivetrain.MIN_SPEED){
```
change to accommodate negative speeds

tuval88 commented on src/main/java/frc/robot/subsystems/SwerveModuleHolder.java:
```diff
@@ -0,0 +1,96 @@
+    private static final String BACK_LEFT_NAMESPACE_NAME = "back left";
+    private static final String BACK_RIGHT_NAMESPACE_NAME = "back right";
+
+    private static final double FRONT_LEFT_OFFSET = -1;
+    private static final double FRONT_RIGHT_OFFSET = -1;
+    private static final double BACK_LEFT_OFFSET = -1;
+    private static final double BACK_RIGHT_OFFSET = -1;
```
use 0 as default value

tuval88 commented on src/main/java/frc/robot/subsystems/SwerveModuleHolder.java:
```diff
@@ -0,0 +1,96 @@
+    private static final double BACK_RIGHT_OFFSET = -1;
+
+    private static final PIDSettings drivePIDSettings = namespace.addPIDNamespace("drive",
+            new PIDSettings(0, 0, 0, 0, 0,0));
```
do not give unnecessary initial values

goooooooooooooooooose replied:
very necessary (for some weird reason)

tuval88 commented on src/main/java/frc/robot/subsystems/SwerveModuleHolder.java:
```diff
@@ -0,0 +1,96 @@
+    private static final PIDSettings drivePIDSettings = namespace.addPIDNamespace("drive",
+            new PIDSettings(0, 0, 0, 0, 0,0));
+    private static final PIDSettings turnPIDSettings = namespace.addPIDNamespace("turn",
+            new PIDSettings(0, 0, 0, 0, 0,0));
```
ditto

tuval88 commented on src/main/java/frc/robot/subsystems/SwerveModuleHolder.java:
```diff
@@ -0,0 +1,96 @@
+    private static final PIDSettings turnPIDSettings = namespace.addPIDNamespace("turn",
+            new PIDSettings(0, 0, 0, 0, 0,0));
+    private static final FeedForwardSettings driveFeedForwardSettings = namespace.addFeedForwardNamespace(
+            "drive", new FeedForwardSettings(0, 0, 0, 0, FeedForwardController.ControlMode.LINEAR_VELOCITY));
```
ditto

tuval88 commented on src/main/java/frc/robot/subsystems/SwerveModuleHolder.java:
```diff
@@ -0,0 +1,96 @@
+    private static final FeedForwardSettings driveFeedForwardSettings = namespace.addFeedForwardNamespace(
+            "drive", new FeedForwardSettings(0, 0, 0, 0, FeedForwardController.ControlMode.LINEAR_VELOCITY));
+    private static final FeedForwardSettings turnFeedForwardSettings = namespace.addFeedForwardNamespace(
+            "turn", new FeedForwardSettings(0, 0, 0, 0, FeedForwardController.ControlMode.LINEAR_POSITION));
```
ditto

goooooooooooooooooose committed:
renamed to rotationSpeed

goooooooooooooooooose committed:
renamed to rotationSpeed

goooooooooooooooooose committed:
merged into 1 line

goooooooooooooooooose committed:
merged into 1 line

goooooooooooooooooose committed:
accommodated minus

goooooooooooooooooose committed:
accommodated minus

goooooooooooooooooose committed:
0 default offset value

goooooooooooooooooose committed:
0 default offset value

goooooooooooooooooose committed:
removed unnecessary values

goooooooooooooooooose committed:
removed unnecessary values

goooooooooooooooooose committed:
ditto

goooooooooooooooooose committed:
ditto

tuval88 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -123,10 +123,7 @@ private SwerveModuleState optimize(SwerveModuleState state, double currentAngle)
-            else {
-                desiredAngle += DEGREES_TO_FLIP;
-            }
+            } else {desiredAngle += DEGREES_TO_FLIP;}
```
unmerge the {}

tuval88 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -16,6 +12,8 @@
 import edu.wpi.first.math.geometry.Rotation2d;
 import edu.wpi.first.math.kinematics.SwerveModuleState;
 
+import static java.lang.Math.abs;
```
why?

goooooooooooooooooose replied:
why not

tuval88 commented on src/main/deploy/pathplanner/navgrid.json:
```diff
@@ -0,0 +1 @@
+{"field_size":{"x":17.548,"y":8.052},"nodeSizeMeters":0.3,"grid":[[true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true],[true,true,true,true,true,true,true, ...
```
delete file

tuval88 commented on src/main/java/frc/robot/subsystems/SwerveModuleHolder.java:
```diff
@@ -37,9 +37,9 @@ public class SwerveModuleHolder {
 
     private static final PIDSettings drivePIDSettings = namespace.addPIDNamespace("drive",
-            new PIDSettings(0, 0, 0, 0, 0,0));
+            new PIDSettings(0, 0, 0));
```
do not give this method settings at all

tuval88 commented on src/main/java/frc/robot/subsystems/SwerveModuleHolder.java:
```diff
@@ -37,9 +37,9 @@ public class SwerveModuleHolder {
+            new PIDSettings(0, 0, 0));
     private static final PIDSettings turnPIDSettings = namespace.addPIDNamespace("turn",
-            new PIDSettings(0, 0, 0, 0, 0,0));
+            new PIDSettings(0, 0, 0));
```
ditto

goooooooooooooooooose committed:
unmerged

goooooooooooooooooose committed:
unmerged

goooooooooooooooooose committed:
deleted unnecessary imports

goooooooooooooooooose committed:
deleted unnecessary imports

goooooooooooooooooose committed:
deleted weird file

goooooooooooooooooose committed:
deleted weird file

goooooooooooooooooose committed:
deleted unnecessary given values

goooooooooooooooooose committed:
deleted unnecessary given values

Spikes-2212 commented on src/main/java/frc/robot/commands/Drive.java:
```diff
@@ -0,0 +1,50 @@
+    public Drive(Drivetrain drivetrain, Supplier<Double> xSpeed, Supplier<Double> ySpeed,
+                 Supplier<Double> rotationSpeed, boolean fieldRelative, boolean usePID) {
+        this.drivetrain = drivetrain;
+        addRequirements(drivetrain);
```
addRequirements() before anything else

Spikes-2212 commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,97 @@
+    public static final double MIN_SPEED = 0.05;
+    public static final double MAX_TURN_SPEED = 3;
+
+    private static final double TRACK_WIDTH = 60;
```
this is in meters, by the way

Spikes-2212 commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,97 @@
+        return instance;
+    }
+
+    private Drivetrain(SwerveModule frontLeft, SwerveModule frontRight, SwerveModule backLeft,
```
call configureDashboard()

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,145 @@
+    private final FeedForwardSettings driveFeedForwardSettings;
+    private final FeedForwardSettings turnFeedForwardSettings;
+
+    public SwerveModule(String namespace, TalonFXWrapper driveMotor, SparkWrapper turnMotor, CANcoder absoluteEncoder,
```
ditto

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,145 @@
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.kinematics.SwerveModuleState;
+
+import static java.lang.Math.abs;
```
![image](https://github.com/user-attachments/assets/67c0d3c6-44f8-4b93-827c-c1e7611ecd4f)

goooooooooooooooooose committed:
add requirements FIRST (get it??? like FIRST the program, do you get it??)

goooooooooooooooooose committed:
add requirements FIRST (get it??? like FIRST the program, do you get it??)

goooooooooooooooooose committed:
changed units to meters

goooooooooooooooooose committed:
changed units to meters

goooooooooooooooooose committed:
called configureDashboard

goooooooooooooooooose committed:
called configureDashboard

goooooooooooooooooose committed:
ditto

goooooooooooooooooose committed:
ditto

goooooooooooooooooose committed:
boom.

goooooooooooooooooose committed:
boom.

TurtleMinecraft commented on src/main/java/frc/robot/RobotMap.java:
```diff
@@ -4,6 +4,20 @@ public class RobotMap {
 
     public interface CAN {
 
+        int FRONT_LEFT_DRIVE_TALON_FX = -1;
+        int FRONT_RIGHT_DRIVE_TALON_FX = -1;
+        int BACK_LEFT_DRIVE_TALON_FX = -1;
+        int BACK_RIGHT_DRIVE_TALON_FX = -1;
+
+        int FRONT_LEFT_TURN_SPARK_MAX = -1;
+        int FRONT_RIGHT_TURN_SPARK_MAX = -1;
+        int BACK_LEFT_TURN_SPARK_MAX = -1;
+        int BACK_RIGHT_TURN_SPARK_MAX = -1;
+
+        int FRONT_LEFT_ABSOLUTE_ENCODER = -1;
+        int FRONT_RIGHT_ABSOLUTE_ENCODER = -1;
+        int BACK_LEFT_ABSOLUTE_ENCODER = -1;
+        int BACK_RIGHT_ABSOLUTE_ENCODER = -1;
```
ports are 1 to 12
front left - talon 1, spark 2, cancoder 3 and so on

TurtleMinecraft commented on src/main/java/frc/robot/commands/Drive.java:
```diff
@@ -0,0 +1,50 @@
+
+import java.util.function.Supplier;
+
+public class Drive extends Command {
```
add an option to disable acceleration limit

TurtleMinecraft commented on vendordeps/SpikesLib2.json:
update to 5.1.1

TurtleMinecraft committed:
updated libraries

TurtleMinecraft committed:
updated libraries

TurtleMinecraft committed:
bruh

TurtleMinecraft committed:
bruh

goooooooooooooooooose committed:
not done yet

goooooooooooooooooose committed:
not done yet

goooooooooooooooooose committed:
updated CAN ports

goooooooooooooooooose committed:
updated CAN ports

goooooooooooooooooose committed:
added reset gyro func

goooooooooooooooooose committed:
added reset gyro func

goooooooooooooooooose committed:
added option to disable acceleration

goooooooooooooooooose committed:
added option to disable acceleration

goooooooooooooooooose committed:
updated SpikesLib2.jsonl

goooooooooooooooooose committed:
updated SpikesLib2.jsonl

TurtleMinecraft commented on vendordeps/SpikesLib2.json:
```diff
@@ -0,0 +1,21 @@
+{
+  "fileName": "SpikesLib2.json",
+  "name": "SpikesLib2",
+  "version": "v5.1.0",
```
5.1.1

goooooooooooooooooose committed:
updated SpikesLib2.json

goooooooooooooooooose committed:
updated SpikesLib2.json

Spikes-2212 commented on src/main/java/frc/robot/commands/Drive.java:
```diff
@@ -0,0 +1,60 @@
+        drivetrain.drive(xSpeed, ySpeed, rotationSpeed, fieldRelative, usePID);
+    }
+
+    public void end(boolean interrupted) {
```
override

Spikes-2212 commented on src/main/java/frc/robot/commands/Drive.java:
```diff
@@ -0,0 +1,60 @@
+    private final SlewRateLimiter xLimiter;
+    private final SlewRateLimiter yLimiter;
+    private final SlewRateLimiter rotationLimiter;
+    private final boolean limitAcceleration;
```
put this below usePID

Spikes-2212 commented on src/main/java/frc/robot/commands/Drive.java:
```diff
@@ -0,0 +1,60 @@
+                 Supplier<Double> rotationSpeed, boolean fieldRelative, boolean usePID,
+                 boolean limitAcceleration) {
+        addRequirements(drivetrain);
+        this.drivetrain = drivetrain;
+        this.xSpeed = xSpeed;
+        this.ySpeed = ySpeed;
+        this.rotationSpeed = rotationSpeed;
+        this.usePID = usePID;
+        this.limitAcceleration = limitAcceleration;
+        this.fieldRelative = fieldRelative;
```
initialize these values in the same order as you receive them in the constructor

Spikes-2212 commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,102 @@
+    private final SwerveModule frontRight;
+    private final SwerveModule backLeft;
+    private final SwerveModule backRight;
+    private final AHRS gyro;
```
line break after this

Spikes-2212 commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,102 @@
+import edu.wpi.first.math.kinematics.SwerveDriveKinematics;
+import edu.wpi.first.math.kinematics.SwerveModuleState;
+
+public class Drivetrain extends DashboardedSubsystem {
```
i think we should try and add odometry

goooooooooooooooooose committed:
added override to drive command end func

goooooooooooooooooose committed:
added override to drive command end func

goooooooooooooooooose committed:
limitAcceleration below usePID

goooooooooooooooooose committed:
limitAcceleration below usePID

goooooooooooooooooose committed:
initializing the values in the same order as received

goooooooooooooooooose committed:
initializing the values in the same order as received

goooooooooooooooooose committed:
added blank line

goooooooooooooooooose committed:
added blank line

goooooooooooooooooose committed:
added odometry

goooooooooooooooooose committed:
added odometry

Spikes-2212 commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,117 @@
+    private final SwerveDriveOdometry odometry;
+    private final SwerveDriveKinematics kinematics;
+
+    private static Pose2d pose2d;
```
currentPose

Spikes-2212 commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,117 @@
+        this.backLeft = backLeft;
+        this.backRight = backRight;
+        this.gyro = gyro;
+        this.pose2d = new Pose2d();
```
no reason to write "this"

Spikes-2212 commented on src/main/java/frc/robot/subsystems/SwerveModule.java:
```diff
@@ -0,0 +1,150 @@
+        turnMotor.setPosition(getAbsoluteAngle());
+    }
+
+    public SwerveModulePosition getSwerveModulePosition() {
```
getModulePosition()

goooooooooooooooooose committed:
renamed currentPose and removed this

goooooooooooooooooose committed:
renamed currentPose and removed this

goooooooooooooooooose committed:
renamed getModulePosition

goooooooooooooooooose committed:
renamed getModulePosition

Spikes-2212 commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,117 @@
+    private final SwerveDriveOdometry odometry;
+    private final SwerveDriveKinematics kinematics;
+
+    private static Pose2d currentPose;
```
okay why is this static

Spikes-2212 commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    @Override
+    public void periodic() {
```
put `periodic()` below the constructor

goooooooooooooooooose committed:
currentPose not static

goooooooooooooooooose committed:
currentPose not static

goooooooooooooooooose committed:
periodic below constructor

goooooooooooooooooose committed:
periodic below constructor

goooooooooooooooooose committed:
added named commands

goooooooooooooooooose committed:
weh

goooooooooooooooooose committed:
weh

goooooooooooooooooose committed:
add empty line

goooooooooooooooooose committed:
below final fields

goooooooooooooooooose committed:
kp is double

goooooooooooooooooose committed:
filenames are the same as their name

goooooooooooooooooose committed:
auto files end with blank line

goooooooooooooooooose committed:
removed blank line

goooooooooooooooooose committed:
ctr alt l

goooooooooooooooooose committed:
changed stupid pointless arrays name

goooooooooooooooooose committed:
correct port to gyro

goooooooooooooooooose committed:
blank line at the end of path files

goooooooooooooooooose committed:
added blank line

goooooooooooooooooose committed:
added blank line

goooooooooooooooooose committed:
changed stupid pointless array name 2.0

goooooooooooooooooose committed:
initialized values correct order

goooooooooooooooooose committed:
ctr alt l

goooooooooooooooooose committed:
Merge branch 'gil-swerve' into gil-path-planner-swerve

goooooooooooooooooose committed:
ctr alt l

goooooooooooooooooose committed:
ctr alt l

goooooooooooooooooose committed:
Merge pull request #7 from Spikes-2212-Programming-Guild/gil-path-planner-swerve

Gil path planner swerve

goooooooooooooooooose committed:
Merge branch 'dev' into gil-swerve

goooooooooooooooooose committed:
safd

goooooooooooooooooose committed:
safd

goooooooooooooooooose committed:
Merge pull request #14 from Spikes-2212-Programming-Guild/gil-path-planner-swerve

Gil path planner swerve
